services/src/main/java/org/keycloak/authorization/protection/policy/PolicyService.java

```diff
@@ -0,0 +1,244 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Response;
+import org.jboss.resteasy.annotations.cache.NoCache;
+import org.keycloak.authorization.AuthorizationProvider;
+import org.keycloak.authorization.identity.Identity;
+import org.keycloak.authorization.model.Policy;
+import org.keycloak.authorization.model.Resource;
+import org.keycloak.authorization.model.ResourceServer;
+import org.keycloak.authorization.store.PolicyStore;
+import org.keycloak.authorization.store.ResourceStore;
+import org.keycloak.models.Constants;
+import org.keycloak.models.utils.ModelToRepresentation;
+import org.keycloak.models.utils.RepresentationToModel;
+import org.keycloak.representations.idm.authorization.UmaPermissionTypeRepresentation;
+import org.keycloak.services.ErrorResponseException;
+
+/**
+ * @author <a href="mailto:federico@martel-innovate.com">Federico M. Facca</a>
+ */
+public class PolicyService {
+
+    private final ResourceServer resourceServer;
+    private final Identity identity;
+    private final AuthorizationProvider authorization;
+    private final ResourceStore resourceStore;
+    private final PolicyStore policyStore;
+
+    public PolicyService(Identity identity, ResourceServer resourceServer, AuthorizationProvider authorization) {
+        this.identity = identity;
+        this.resourceServer = resourceServer;
+        this.authorization = authorization;
+        this.resourceStore = this.authorization.getStoreFactory().getResourceStore();
+        this.policyStore = this.authorization.getStoreFactory().getPolicyStore();
+    }
+
+    @POST
+    @Path("{resourceId}")
+    @Consumes("application/json")
+    @Produces("application/json")
+    public Response create(@PathParam("resourceId") String resourceId, UmaPermissionTypeRepresentation policy) {
+        validateRequest(resourceId);
+        validateUmaPermissionTypeRepresentation(resourceId, false, policy);
+        validateScopes(resourceId, policy.getScopes());
+        Policy model = RepresentationToModel.createPolicyfromUmaPermissionTypeRepresentation(policy, resourceId, resourceServer, authorization, identity);
+        return Response.ok().entity(ModelToRepresentation.toUmaPermissionTypeRepresentation(model, authorization)).build();
+    }
+
+    @Path("{resourceId}/{policyId}")
+    @PUT
+    @Consumes("application/json")
+    @Produces("application/json")
+    public Response update(@PathParam("resourceId") String resourceId, @PathParam("policyId") String policyId, UmaPermissionTypeRepresentation policy) {
```

Comment:
if i recall correctly from the gdoc, we agreed that for flexibility a resource could have more "permissions" associated. so that each policy in the end is evaluated like an "or".

Reply:
In this case, we can just update the "decision strategy" of the uma permission to "affirmative" as ti will works like an "or".

Reply:
correct, but then either we support (which we agreed to remove) aggregated policies, or we won't be able to say things as:

authorise if

1. permission1 user role A

  or

1. permission2 user  role B and group Z

Reply:
That is fine because you are creating a ```uma``` permission type. This type of policies should be associated with multiple "policies". To achieve the "or" behavior you just set the decision strategy to ```AFFIRMATIVE``` when creating the "uma" permission type.

In other words, you don't need to manage multiple permissions for a single resource, but just a single one by updating the scopes and associated policies.

Reply:
mmm, i am confused now...

I tested the behaviour of your code.

1. If user `A` ask for `READ` permission on resource `Z` of user `B`.  A UMA permission is created attached to resource `Z`.
1. If user `C` ask for `WRITE` permission on resource `Z` of user `B`.  An additional UMA permission is created attached to resource `Z`.

so if i want to see, remove, or edit only one of the two created `uma` permissions for resource `Z` i need the id that identify such permission.

also, if we would have a single uma permission per resource, how we would be able to assign different scopes for the same resource to different users?

Reply:
My changes are all about the "normal" UMA flow, where a requester is asking access to the owner of a resource.So, yeah, we have different UMA permissions for Resource Z each one specific for User A and User C. It is easier to identify those permissions, as you just need to know the resource and the requester, so you can obtain a ticket which is then associated with the UMA permission.

That is what I have commented here about adding the requester to your ```/uma-policy```, so you could manage a specific policy for a specific requester/user. But that won't work for you, as you need to define additional policies to a resource that are not necessarily linked to a requester. For instance, if you define group or role policies.

Now think about this scenario. You have created a UMA permission using ```/uma-policy``` granting access to a specific role or group. Now you want to check if a user has access to the resource associated with this permission. As you know, you can ask for permissions using a ticket (UMA flow), using a permission request and passing the resource(s) or just obtain all permissions from the server. The two first methods are OK since you know the resource you need to check for permissions. But what about the last method ? Obtain all permissions from the server ? How we would perform this evaluation without iterating over all existing resources in the server ? That may impact performance ....

Reply:
1. we can add requester for ticket permission generated UMA permissions as query parameter.
1. we can use in a "tricky" way requester for the other policies. e.g. if it is not a uma permission, requester is used to find all the policy where related to "requesterId" via user, group or role.

Reply:
I think I understand #1. And that would allow users to manage their policies created using UMA flow through a permission ticket (or sharing via Account Services).

But did not understand #2.

Reply:
i proposed that to have a quick way from a ui to see which resources "using" uma a user is related to so that they can be displayed in the ui without obtaining all permissions.